cse/indexing/DocumentMapBuilder.py

```diff
@@ -0,0 +1,105 @@
+            try:
+                os.makedirs(os.path.dirname(document_map_index))
+            except OSError as exc: # Guard against race condition
+                if exc.errno != errno.EEXIST:
+                    raise
+
+        self.__document_map_index = document_map_index
+        self.__document_map_dict = document_map_dict
+        self.__index = []
+        self.__i = 0
+        self.__content = 0
+        if os.path.exists(self.__document_map_index):
+            os.remove(self.__document_map_index)
+        if os.path.exists(self.__document_map_dict):
+            os.remove(self.__document_map_dict)
+        dir = os.path.dirname(self.__document_map_index)
+        if os.path.isdir(os.path.join(dir, 'document_map_partials')):
+            shutil.rmtree(os.path.join(dir, 'document_map_partials'))
+
+    def close(self):
+        self.__index.sort()
+        self.__save_partial_index(self.__index, self.__i)
+        doc_dict = self._build_index_and_dict()
+        PackerUtil.packToFile(doc_dict, self.__document_map_dict, type=PackerUtil.PICKLE)
+
+    def insert(self, cid, pointer):
+        self.__index.append((cid, pointer))
+        self.__content += 1
+        if self.__content >= PARTIAL_THRESHOLD:
+            self.__index.sort()
+            self.__save_partial_index(self.__index, self.__i)
+            self.__index = []
+            self.__content = 0
+            self.__i += 1
+
+    def get(self, cid):
+        return self.__index[cid]
+
+    def listCids(self):
+        return [cid for cid in self.__index]
+
+    def __enter__(self):
+        return self.open()
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def __save_partial_index(self, partial_list, i):
+        dir = os.path.dirname(self.__document_map_index)
+        file_name = os.path.basename(self.__document_map_index)
+        if not os.path.exists(os.path.join(dir, 'document_map_partials')):
+            os.makedirs(os.path.join(dir, 'document_map_partials'))
+
+        with open(os.path.join(dir, 'document_map_partials', file_name + str(i)), 'w') as partial:
+            for tuple in partial_list:
+                partial.write(str(tuple[0]) + ',' + str(tuple[1]) + '\n')
+
+    def _build_index_and_dict(self):
+        doc_dict = ([], [])
+        dir = os.path.dirname(self.__document_map_index)
+        partials_dir = os.path.join(dir, 'document_map_partials')
+        partial_files = []
+        current_lines = []
+        skips = SKIP_SIZE
+        with open(self.__document_map_index, 'w') as out_index:
+            for file in os.listdir(partials_dir):
+                partial_files.append(open(os.path.join(partials_dir, file)))
+
+            for file_handle in partial_files:
+                current_lines.append(self._read_partial_line(file_handle))
+
+            while partial_files:
+                min_index = current_lines.index(min(current_lines))
+                skips += 1
+                if skips >= SKIP_SIZE:
```

Comment:
this is the chunk size for which a pointer is stored in the in-memory dictionary, isn't it?

Reply:
right